### PR TITLE
mold: init at 0.9.2

### LINKED ISSUE #128706: package request: mold
**Project description**
> mold is a high performance drop-in replacement for existing Unix linkers. It is several times faster than LLVM lld linker, the (then-) fastest open-source linker which [they] originally created a few years ago.

**Metadata**

* homepage URL: https://github.com/rui314/mold
* source URL: https://github.com/rui314/mold
* license: GNU AGPL v3
* platforms: unix, linux, darwin

### PR DESCRIPTION
**Pull request by nitsky**
###### Motivation for this change

Add [mold](https://github.com/rui314/mold), a fast linker for x86_64 linux. Fixes #128706.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by legendofmiracles** [comment]
Missing `meta` section

File: pkgs/development/tools/mold/default.nix
Comment (legendofmiracles) [inline pkgs/development/tools/mold/default.nix:5]: use fetchFromGitHub instead and don't hardcode the version, but instead use the `version` variable

File: pkgs/development/tools/mold/default.nix
Comment (legendofmiracles) [inline pkgs/development/tools/mold/default.nix:2]:
```suggestion

stdenv.mkDerivation rec {
```

**Comment by nitsky** [comment]
@legendofmiracles I addressed your comments. Thank you for the review!

File: pkgs/development/tools/mold/default.nix
Comment (piegamesde) [inline pkgs/development/tools/mold/default.nix:12]: This line looks weird like this.
```suggestion
  dontUseCmakeConfigure = true;
```

**Comment by SCOTT-HAMILTON** [comment]
If you wan't it to use lld, you must add 
```nix
{
  EXTRA_LDFLAGS = "--ld-path=${lld_12}/bin/ld.lld";
}
```
I tried with LTO on and it doesn't seem too long :
```nix
{
  LTO = 1;
}
```

File: pkgs/development/tools/mold/default.nix
Comment (SCOTT-HAMILTON) [inline pkgs/development/tools/mold/default.nix:18]: We should ask wether this is agpl3Only or agpl3Plus according to https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206

File: pkgs/development/tools/mold/default.nix
Comment (SCOTT-HAMILTON) [inline pkgs/development/tools/mold/default.nix:11]: tbb, zlib and openssl are build time library dependencies, they should probably be moved to buildInputs.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 128889` at d404d895 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>mold</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/tools/mold/default.nix:14:3:
```
   |
14 |   installPhase = "mkdir -p $out $out/bin $out/share/man/man1 && PREFIX=$out make install";
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`agpl3` is a deprecated license, please check if project uses `agpl3Plus` or `agpl3Only` and change `meta.license` accordingly.

Near pkgs/development/tools/mold/default.nix:18:5:
```
   |
18 |     license = lib.licenses.agpl3;
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/tools/mold/default.nix:13:3:
```
   |
13 |   buildPhase = "make -j $NIX_BUILD_CORES";
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/tools/mold/default.nix:15:3:
```
   |
15 |   meta = {
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/development/tools/mold/default.nix
Comment (nitsky) [reply pkgs/development/tools/mold/default.nix:13]: On my machine, without it only one CPU core is used.

File: pkgs/development/tools/mold/default.nix
Comment (nitsky) [reply pkgs/development/tools/mold/default.nix:16]: I changed it to "A high performance drop-in replacement for existing unix linkers", including the instruction at the link you sent to omit the name of the package.

File: pkgs/development/tools/mold/default.nix
Comment (nitsky) [reply pkgs/development/tools/mold/default.nix:11]: tbb, zlib and openssl are ~~build~~ *run* time library dependencies. I moved them to buildInputs, thanks.

File: pkgs/development/tools/mold/default.nix
Comment (nitsky) [reply pkgs/development/tools/mold/default.nix:13]: I removed the custom `buildPhase` entirely, and I find that still only one CPU core is used.

**Comment by nitsky** [comment]
@SCOTT-HAMILTON I adopted all of your suggestions. Thanks!

**Review by SCOTT-HAMILTON — COMMENTED** [review]
LGTM

**Comment by legendofmiracles** [comment]
Missing maintainer, and it's still quite unreadable - consider adding a newlines between different sections of the derivation

**Comment by nitsky** [comment]
@legendofmiracles should I add myself to the maintainers list in this commit, or in a separate commit?

**Comment by nitsky** [comment]
@piegamesde thanks, submitted #129030.

**Comment by nitsky** [comment]
@legendofmiracles I submitted a PR to add myself to the maintainers, and updated this PR with the maintainer and added spacing.

**Review by legendofmiracles — APPROVED** [review]
Meta section usually goes to the end of the file, but otherwise it looks good to me

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:1]:
```suggestion
{ stdenv, fetchFromGitHub, lib, autoPatchelfHook, cmake, tbb, xxHash, zlib, openssl }:
```

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:5]: Could you move meta section to end of expression? This will increase readability.

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:20]:
```suggestion
  nativeBuildInputs = [ autoPatchelfHook cmake xxHash ];
```

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:23]:
```suggestion

```

File: pkgs/top-level/all-packages.nix
Comment (musfay) [inline pkgs/top-level/all-packages.nix:13817]:
```suggestion
  mold = callPackage ../development/tools/mold { stdenv = llvmPackages_latest.stdenv; };
```

**Review by musfay — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:23]:
```suggestion
  EXTRA_LDFLAGS = "--ld-path=${llvmPackages_latest.lld}/bin/ld.lld";
```

**Review by musfay — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:1]:
```suggestion
{ stdenv, fetchFromGitHub, lib, autoPatchelfHook, cmake, tbb, llvmPackages_latest, xxHash, zlib, openssl }:
```

**Review by musfay — CHANGES_REQUESTED** [review]

**Comment by nitsky** [comment]
@musfay thank you for the suggestions I adopted all of them. I also added a commit adding myself to the maintainers list.

**Review by legendofmiracles — APPROVED** [review]

File: pkgs/development/tools/mold/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:26]: Why do we want a link to bin? It is a shared object, not a binary, right?

File: pkgs/development/tools/mold/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:2]:
```suggestion
{ stdenv, fetchFromGitHub, lib, autoPatchelfHook, cmake, tbb, llvmPackages_latest, xxHash, zlib, openssl }:

stdenv.mkDerivation rec {
```

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:13817]:
```suggestion
  mold = callPackage ../development/tools/mold {
    stdenv = llvmPackages_latest.stdenv;
  };
```

File: pkgs/development/tools/mold/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:33]:
```suggestion
  meta = with lib; {
    description = "A high performance drop-in replacement for existing unix linkers";
    homepage = "https://github.com/rui314/mold";
    license = licenses.agpl3Plus;
    maintainers = with maintainers; [ nitsky ];
```

File: pkgs/development/tools/mold/default.nix
Comment (piegamesde) [reply pkgs/development/tools/mold/default.nix:2]: The line is long enough. How about:

```nix
{
  stdenv,
  fetchFromGitHub,
  …
  openssl,
}:
```
?

File: pkgs/development/tools/mold/default.nix
Comment (nitsky) [reply pkgs/development/tools/mold/default.nix:26]: @SCOTT-HAMILTON said:

> You should also probably add a symbolic link of $out/lib/mold/mold-wrapper.so to $out/bin because of https://github.com/rui314/mold/blob/9a09c777d9460ebe7eb498d1cc0327915f8bbcdc/subprocess.cc#L271

File: pkgs/development/tools/mold/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/mold/default.nix]: If you want but put the ``,`` at the beginning of the line.

File: pkgs/development/tools/mold/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:12]:
```suggestion
{ stdenv
, fetchFromGitHub
, lib
, autoPatchelfHook
, cmake
, tbb
, llvmPackages_latest
, xxHash
, zlib
, openssl
}:
```

File: pkgs/development/tools/mold/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/mold/default.nix:38]: The bin directory is for executables that you can run in your shell. I'd would rather see this path patched instead.

File: pkgs/development/tools/mold/default.nix
Comment (piegamesde) [reply pkgs/development/tools/mold/default.nix:12]: As long as we don't have an official style guide that specifies the preferred formatting in this situation, I call this opinionated.

Comment (SuperSandro2000) [reply pkgs/development/tools/mold/default.nix:12]: That is not opinionated, this is the default style for nixpkgs.

**Comment by piegamesde** [comment]
@rui314 can you please review this?

File: pkgs/development/tools/mold/default.nix
Comment (rui314) [reply pkgs/development/tools/mold/default.nix:37]: You can remove a symlink in 0.9.2, as mold installed as $out/bin/mold checks if $out/lib/mold-wrapper.so exists whatever $bin is. But 0.9.1 unfortunately tries to do that only if $bin is /usr or /usr/local. So, I think we need it for now.

Comment (SuperSandro2000) [reply pkgs/development/tools/mold/default.nix:37]: You can also fetch the commit it was changed as a patch.

**Comment by legendofmiracles** [comment]
Umm, why?

**Comment by nitsky** [comment]
@legendofmiracles I mistakenly deleted the branch while attempting to rebase. Can this PR be reopened?

**Comment by nitsky** [comment]
@legendofmiracles I rebased and resolved the outstanding issues brought up in reviews.

File: pkgs/development/tools/mold/default.nix
Comment (nitsky) [reply pkgs/development/tools/mold/default.nix]: I updated this PR to use 0.9.2, released in the last few hours, and deleted this symlink.

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:34]:
```suggestion

```

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:19]:
```suggestion
    repo = pname;
```

**Review by musfay — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [reply pkgs/development/tools/mold/default.nix:34]: Fixed with 0.9.2

Comment (nitsky) [reply pkgs/development/tools/mold/default.nix]: Thanks!

File: pkgs/development/tools/mold/default.nix
Comment (musfay) [inline pkgs/development/tools/mold/default.nix:28]:
```suggestion
  EXTRA_LDFLAGS = "-fuse-ld=${llvmPackages_latest.lld}/bin/ld.lld";
```
Isn't it the proper way to use custom linker?

Comment (nitsky) [reply pkgs/development/tools/mold/default.nix:28]: I'm not sure, I just adopted this as a suggestion from @SCOTT-HAMILTON

Comment (musfay) [reply pkgs/development/tools/mold/default.nix:28]: [lld page](https://lld.llvm.org/) also suggests this way.

**Comment by rui314** [comment]
LGTM

File: pkgs/development/tools/mold/default.nix
Comment (nitsky) [reply pkgs/development/tools/mold/default.nix]: I adopted the suggestion, thanks.

**Comment by nitsky** [comment]
Is there anything blocking this PR at this point?

**Review by piegamesde — APPROVED** [review]

**Comment by nitsky** [comment]
@SuperSandro2000 can you approve this PR as well?

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 can you approve this PR as well?

Approving a PR does not really do anything.

**Comment by nitsky** [comment]
> Approving a PR does not really do anything.

Okay, then what are the steps to get this PR merged?

File: pkgs/development/tools/mold/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/mold/default.nix:36]:
```suggestion
    maintainers = with maintainers; [ nitsky ];
    broken = stdenv.isAarch64;
```

I was about the merge it when I noticed that aarch64 builds fail.

**Comment by SuperSandro2000** [comment]
> Okay, then what are the steps to get this PR merged?

I just marked it broken on aarch64. I will wait for ofborg and then merge it.

**Comment by jakobrs** [comment]
I noticed that `mold --run` works, but isn't "effective", on NixOS. `mold --run` works by LD_PRELOADing a library that intercepts certain functions (`execlp`, `execvpe`, `execvp`, `execv`, `posix_spawn`, `execle`, `execve`, and `execl`) and checking if you're trying to launch `/usr/bin/ld{,.lld,.gold}`, and if you are, launches mold instead. The problem is that `/usr/bin` isn't in path on NixOS by default. A solution would be to add a directory containing symlinks to `$out/bin/mold` to PATH.

Some notes (because GitHub still doesn't support footnotes):
- A list of intercepted functions can be generated using `readelf -s result/lib/mold/mold-wrapper.so`.
- `mold --run` is the recommended way to use Mold with Cargo, which is why this matters.
- The source code of mold-wrapper.so is [here](https://github.com/rui314/mold/blob/main/mold-wrapper.c), and the source code of `mold --run` is [here](https://github.com/rui314/mold/blob/main/subprocess.cc#L276-L303)

**Comment by rui314** [comment]
@jakobrs We can relax the rule to recognize commands as "ld". If we want, we can just treat `ld`, `ld.lld` and `ld.gold` as an ld command, regardless of their location in the directory hierarchy.

**Comment by jakobrs** [comment]
@rui314 I tried to do this, but notice how `ld` never gets execed during `cargo build`.

image

gcc works if I change it to use the basename, though.

**Comment by rui314** [comment]
@jakobrs I created a bug for mold to keep track of the issue: https://github.com/rui314/mold/issues/101

**Comment by jakobrs** [comment]
I bumped the version of mold, applied the patch, and created a PR: #131830

**Pull request merged**
